// template-lint/dist/linter.js
"use strict";
const stream_1 = require("stream");
const parser_builder_1 = require("./parser-builder");
const ast_1 = require("./ast");
class Linter {
    constructor(rules, parserBuilder, ast) {
        this.rules = rules || [];
        this.parserBuilder = parserBuilder || new parser_builder_1.ParserBuilder();
        this.ast = ast || new ast_1.AST();
    }
    lint(html, path) {
        var parser = this.parserBuilder.build();
        parser.init(this.rules, this.ast, path);
        if (typeof (html) === 'string') {
            var stream = new stream_1.Readable();
            stream.push(html);
            stream.push(null);
            stream.pipe(parser);
        }
        else if (this.isStream(html)) {
            html.pipe(parser);
        }
        else {
            throw new Error("html isn't pipeable");
        }
        var completed = new Promise(function (resolve, reject) {
            parser.on("end", () => {
                parser.finalise();
                resolve();
            });
        });
        var ruleTasks = [];
        this.rules.forEach((rule) => {
            let task = completed.then(() => {
                return rule.finalise(this.ast.root);
            });
            ruleTasks.push(task);
        });
        return Promise.all(ruleTasks).then(results => {
            var all = new Array();
            results.forEach(parts => {
                all = all.concat(parts);
            });
            all = all.sort((a, b) => (a.line - b.line) * 1000 + (a.column - b.column));
            return all;
        });
    }
    isStream(input) {
        return input.pipe && typeof (input.pipe) === "function";
    }
}
exports.Linter = Linter;
