// aurelia-template-lint/dist/aurelia-reflection.js
"use strict";
const aurelia_templating_binding_1 = require("aurelia-templating-binding");
const aurelia_templating_1 = require("aurelia-templating");
const aurelia_dependency_injection_1 = require("aurelia-dependency-injection");
class AureliaReflection {
    constructor() {
        this.container = new aurelia_dependency_injection_1.Container();
        this.resources = this.container.get(aurelia_templating_1.ViewResources);
        this.bindingLanguage = this.container.get(aurelia_templating_binding_1.TemplatingBindingLanguage);
    }
    createAttributeInstruction(tag, name, value) {
        var instruction = null;
        let info = this.bindingLanguage.inspectAttribute(this.resources, tag, name, value);
        if (info)
            instruction = this.bindingLanguage.createAttributeInstruction(this.resources, { tagName: tag }, info, undefined);
        return instruction;
    }
    createTextExpression(text) {
        var exp = null;
        exp = this.bindingLanguage.inspectTextContent(this.resources, text);
        return exp;
    }
}
exports.AureliaReflection = AureliaReflection;
