// aurelia-template-lint/dist/aurelia-linter.js
"use strict";
const template_lint_1 = require('template-lint');
const template_lint_2 = require('template-lint');
const template_lint_3 = require('template-lint');
const template_lint_4 = require('template-lint');
const template_lint_5 = require('template-lint');
const template_lint_6 = require('template-lint');
const template_lint_7 = require('template-lint');
const template_lint_8 = require('template-lint');
const template_lint_9 = require('template-lint');
const template_lint_10 = require('template-lint');
const require_1 = require('./rules/require');
const slot_1 = require('./rules/slot');
const template_1 = require('./rules/template');
const binding_1 = require('./rules/binding');
const required_attr_1 = require('./rules/required-attr');
const reflection_1 = require('./reflection');
const aurelia_reflection_1 = require('./aurelia-reflection');
const config_1 = require('./config');
const aurelia_pal_nodejs_1 = require('aurelia-pal-nodejs');
aurelia_pal_nodejs_1.initialize();
class AureliaLinter {
    constructor(config) {
        if (config == undefined)
            config = new config_1.Config();
        this.config = config;
        this.reflection = new reflection_1.Reflection();
        this.auReflection = new aurelia_reflection_1.AureliaReflection();
        let rules = [];
        if (this.config.useRuleSelfClose)
            rules.push(new template_lint_3.SelfCloseRule());
        if (this.config.useRuleStructure)
            rules.push(new template_lint_4.StructureRule());
        if (this.config.useRuleObsoleteAttribute)
            rules.push(new template_lint_6.ObsoleteAttributeRule(config.obsoleteAttributeOpts));
        if (this.config.useRuleObsoleteTag)
            rules.push(new template_lint_5.ObsoleteTagRule(config.obsoleteTagOpts));
        if (this.config.useRuleAttributeValue)
            rules.push(new template_lint_8.AttributeValueRule(config.attributeValueOpts));
        if (this.config.useRuleConflictingAttribute)
            rules.push(new template_lint_10.ConflictingAttributesRule(config.conflictingAttributeOpts));
        if (this.config.useRuleId)
            rules.push(new template_lint_7.IdAttributeRule(this.config.idAttributeOpts));
        if (this.config.useRuleValidChildren)
            rules.push(new template_lint_9.ValidChildRule(this.config.validChildOpts));
        if (this.config.useRuleRequiredAttributes)
            rules.push(new required_attr_1.RequiredAttributeRule(this.config.requiredAttribute));
        if (this.config.useRuleAureliaRequire)
            rules.push(new require_1.RequireRule());
        if (this.config.useRuleAureliaSlot)
            rules.push(new slot_1.SlotRule(config.aureliaSlotOpts.controllers));
        if (this.config.useRuleAureliaTemplate)
            rules.push(new template_1.TemplateRule(config.aureliaTemplateOpt.containers));
        if (this.config.useRuleAureliaBindingAccess || this.config.useRuleAureliaBindingSyntax)
            rules.push(new binding_1.BindingRule(this.reflection, this.auReflection, {
                reportBindingSyntax: this.config.useRuleAureliaBindingSyntax,
                reportBindingAccess: this.config.useRuleAureliaBindingAccess,
                reportUnresolvedViewModel: this.config.aureliaBindingAccessOpts.reportUnresolvedViewModel,
                reportExceptions: this.config.debug,
                localOverride: this.config.aureliaBindingAccessOpts.localOverride,
                localProvidors: this.config.aureliaBindingAccessOpts.localProvidors,
                restrictedAccess: this.config.aureliaBindingAccessOpts.restrictedAccess
            }));
        if (this.config.customRules)
            rules.concat(config.customRules);
        var parserBuilder = new template_lint_2.ParserBuilder()
            .withVoids(this.config.parserOpts.voids)
            .withScopes(this.config.parserOpts.scopes);
        this.linter = new template_lint_1.Linter(rules, parserBuilder);
        this.init = this.processSourceGlobs(this.config.reflectionOpts.sourceFileGlob)
            .then(() => {
            if (this.config.reflectionOpts.typingsFileGlob != null)
                return this.processTypingsGlobs(this.config.reflectionOpts.typingsFileGlob);
            else
                return Promise.resolve();
        });
    }
    lint(html, path) {
        if (this.init) {
            return this.init.then(() => {
                this.init = null;
                return this.linter.lint(html, path);
            });
        }
        else {
            return this.linter.lint(html, path);
        }
    }
    processSourceGlobs(globs) {
        if (typeof (globs) == "string") {
            return this.reflection.addGlob(globs);
        }
        else {
            return Promise.all(globs.map(x => this.reflection.addGlob(x)));
        }
    }
    processTypingsGlobs(globs) {
        if (typeof (globs) == "string") {
            return this.reflection.addTypingsGlob(globs);
        }
        else {
            return Promise.all(globs.map(x => this.reflection.addTypingsGlob(x)));
        }
    }
}
exports.AureliaLinter = AureliaLinter;
