// aurelia-template-lint/dist/rules/required-attr.js
"use strict";
const template_lint_1 = require('template-lint');
const template_lint_2 = require('template-lint');
/**
 * Rule to ensure elements have required attributes
 */
class RequiredAttributeRule extends template_lint_1.Rule {
    constructor(patterns) {
        super();
        this.patterns = patterns ? patterns : [];
    }
    init(parser) {
        parser.on("startTag", (tag, attrs, selfClosing, loc) => {
            let rules = this.patterns.filter(r => {
                return r.tag.test(tag);
            });
            if (rules) {
                for (var rule of rules) {
                    if (!attrs.find(x => rule.attr.test(x.name))) {
                        let issue = new template_lint_2.Issue({
                            message: rule.msg,
                            severity: template_lint_2.IssueSeverity.Error,
                            line: loc.line,
                            column: loc.col,
                            start: loc.startOffset,
                            end: loc.endOffset
                        });
                        this.reportIssue(issue);
                    }
                }
            }
        });
    }
}
exports.RequiredAttributeRule = RequiredAttributeRule;
;
